packages/web-components/fast-foundation/src/listbox/listbox.ts

```diff
@@ -0,0 +1,338 @@
+
+        const selectedIndex = this.listboxItems.findIndex(el => el.isEqualNode(captured));
+        if (selectedIndex !== -1) {
+            this.selectedOptionIndex = selectedIndex;
+            this.selectedOption = this.listboxItems[selectedIndex];
+        }
+    };
+
+    /**
+     * Move focus to an option whose label matches characters typed by the user.
+     * Consecutive keystrokes are batched into a buffer of search text used
+     * to match against the set of options.  If TYPE_AHEAD_TIMEOUT_MS passes
+     * between consecutive keystrokes, the search restarts.
+     *
+     * @param typedKey - the key to be evaluated
+     */
+    private typeAheadValue: string = "";
+    private typeAheadTimeoutHandler: number = -1;
+    private typeAheadExpired: boolean = false;
+    private static readonly TYPE_AHEAD_TIMEOUT_MS = 1000;
+    public handleTypeAhead(typedKey) {
+        // For every keystroke, reset the timer that triggers when enough
+        // time has elapsed such that the search should restart.
+        window.clearTimeout(this.typeAheadTimeoutHandler);
+        this.typeAheadTimeoutHandler = window.setTimeout(() => {
+            this.typeAheadExpired = true;
+        }, Listbox.TYPE_AHEAD_TIMEOUT_MS);
+
+        if (this.typeAheadExpired) {
+            this.typeAheadValue = "";
+        }
+
+        this.typeAheadValue = `${this.typeAheadValue}${typedKey}`;
+
+        const focusedIndex = this.listboxItems.indexOf(document.activeElement as any);
+        const searchStartOffset = this.typeAheadExpired ? 1 : 0;
+        // Try to match first against options that come after the currently
+        // selected option. If none of those match, loop back around starting
+        // from the top of the list. If we're in the middle of a search,
+        // continue matching against the currently focused option before moving
+        // on to the next option.
+        const optionsForSearch = this.listboxItems
+            .slice(focusedIndex + searchStartOffset, this.listboxItems.length)
+            .concat(this.listboxItems.slice(0, focusedIndex + searchStartOffset));
+
+        const pattern = `^(${this.typeAheadValue.replace(
+            /[-/\\^$*+?.()|[\]{}]/g,
+            "\\$&"
+        )})`;
+        const re = new RegExp(pattern, "gi");
+
+        for (const option of optionsForSearch) {
+            // Match against the visible text of the option, rather than
+            // the `value` attribute. For a real <option> element, the
+            // `label` property could be used here.
+            // Chromium/Firefox's native <select>s ignore whitespace at the
+            // beginning/end of the visible text when matching, so trim()
+            // the search text to align with that behavior.
+            const element: any = option;
+            const matches = element.innerText.trim().match(re);
+
+            if (matches) {
+                // this.setFocusOnOption(element);
+                console.log(element);
```

Comment:
does this area still need work?

Reply:
This entire `handleTypeAhead` function is left mostly as-is from the original draft implementation. It needs to be fixed and cleaned up.